### PR TITLE
Rename `CatalogList` to `CatalogProviderList`

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Closes #.

## Rationale for this change
A follow up on https://github.com/apache/arrow-datafusion/pull/8968/files#r1464114449

 Rename `CatalogList` to `CatalogProviderList` as it better reflects the trait meaning which in fact is a collection of catalog providers.

## Are these changes tested?
Yes

## Are there any user-facing changes?
Downstream users should be using a new name `CatalogProviderList` instead of `CatalogList`

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thank you @comphead  -- I think this is a good change.

I wonder if we could make this backwards compatible like

```rust
[deprecated("use CatalogProviderList")]
trait CatalogList : CatalogProviderList {
}
```

Though that would probably still require people to update the code, but maybe it would help them to more easily figure out what they needed to change 🤔

**Comment by comphead** [comment]
> Thank you @comphead -- I think this is a good change.
> 
> I wonder if we could make this backwards compatible like
> 
> ```rust
> [deprecated("use CatalogProviderList")]
> trait CatalogList : CatalogProviderList {
> }
> ```
> 
> Though that would probably still require people to update the code, but maybe it would help them to more easily figure out what they needed to change 🤔

Thats damn good idea.

File: datafusion/core/src/catalog/mod.rs [hunk 1]
```diff
@@ -53,28 +53,32 @@ pub trait CatalogList: Sync + Send {
     fn catalog(&self, name: &str) -> Option<Arc<dyn CatalogProvider>>;
 }
 
+/// See [`CatalogProviderList`]
```
Comment (alamb) [inline datafusion/core/src/catalog/mod.rs:56]: 👍  -- this will probably still error on compilation, but at least it should point users in the right way 👍

**Review by alamb — APPROVED** [review]

**Comment by comphead** [comment]
Thanks @alamb

**Pull request merged**